execution_engine/src/system/handle_payment/internal.rs

```diff
@@ -106,81 +107,89 @@ pub(crate) fn finalize_payment<P: MintProvider + RuntimeProvider + StorageProvid
+                            Ok(()) => {}
+                            Err(error) => {
+                                error!(
+                                    %error,
+                                    %refund_amount,
+                                    %account,
+                                    "unable to transfer refund to a refund purse; refunding to account"
+                                );
+                                refund_to_account::<P>(
+                                    provider,
+                                    payment_purse,
+                                    account,
+                                    refund_amount,
+                                )?;
+                            }
+                        }
+                    }
+                    None => {
+                        refund_to_account::<P>(provider, payment_purse, account, refund_amount)?;
+                    }
+                }
+            }
+        }
+
+        RefundHandling::Burn { .. } if !refund_amount.is_zero() => {
+            // Fee-handling is set to `Burn`.  Deduct the refund from the payment purse and
+            // reduce the total supply (i.e. burn the refund).
             payment_amount = payment_amount
                 .checked_sub(refund_amount)
                 .ok_or(Error::ArithmeticOverflow)?;
 
             provider.write_balance(payment_purse, payment_amount)?;
             provider.reduce_total_supply(refund_amount)?;
-        } else {
-            match refund_purse {
-                Some(refund_purse) => {
-                    // In case of failure to transfer to refund purse we fall back on the account's
-                    // main purse
-                    match provider.transfer_purse_to_purse(
-                        payment_purse,
-                        refund_purse,
-                        refund_amount,
-                    ) {
-                        Ok(()) => {}
-                        Err(error) => {
-                            error!(
-                                %error,
-                                %refund_amount,
-                                %account,
-                                "unable to transfer refund to a refund purse; refunding to account"
-                            );
-                            refund_to_account::<P>(
-                                provider,
-                                payment_purse,
-                                account,
-                                refund_amount,
-                            )?;
-                        }
-                    }
-                }
-                None => {
-                    refund_to_account::<P>(provider, payment_purse, account, refund_amount)?;
-                }
-            }
+        }
+
+        RefundHandling::Burn { .. } => {
+            // No refund to burn
         }
     }
 
-    // Pay or burn the reward.
+    // Pay or burn the fee.
     if target == URef::default() {
```

Comment:
Just realised it would be a lot more explicit if we did `match provider.fee_handling()` here.

Reply:
Did this as well (with debug_assert to ensure `target == URef::default()` to avoid calling it with any concrete purse)